walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,89 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.ArrayList;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
-}
+
+    /** Check if the event and the request contains the same attendee. */
+    private boolean containSameAttendees(Collection < String > eventAttendees, Collection < String > requestAttendees) {
+        if (requestAttendees.size() == 0)
+            return false;
+
+        for (String attendee: eventAttendees)
+            if (requestAttendees.contains(attendee))
+                return true;
+        return false;
+    }
+
+    /**
+     * Combine overlapped TimeRanges in a sorted List, return collection without overlap
+     */
+    private Collection < TimeRange > combineTimeRanges(List < TimeRange > timeRanges) {
+        Collection < TimeRange > combinedTimeRanges = new ArrayList < > ();
+
+        if (timeRanges.size() == 0)
+            return combinedTimeRanges;
+
+        TimeRange currentEvent = timeRanges.get(0);
+        for (int i = 1; i < timeRanges.size(); i++) {
+            TimeRange nextEvent = timeRanges.get(i);
+            if (currentEvent.overlaps(nextEvent)) {
+                int start = currentEvent.start();
+                int end = currentEvent.contains(nextEvent.end()) ? currentEvent.end() : nextEvent.end();
+                currentEvent = TimeRange.fromStartEnd(start, end, false);
+            } else {
+                combinedTimeRanges.add(currentEvent);
+                currentEvent = nextEvent;
+            }
+        }
+        combinedTimeRanges.add(currentEvent);
+
+        return combinedTimeRanges;
+    }
+
+    /**
+     * Get a Collection of free TimeRanges given a collection of busy TimeRanges.
+     */
+    private Collection < TimeRange > getFreeTimeRanges(Collection < TimeRange > busyTimeRanges, Long requestDuration) {
+        Collection < TimeRange > freeTimeRanges = new ArrayList < > ();
+        int start = TimeRange.START_OF_DAY;
+        for (TimeRange time: busyTimeRanges) {
+            int newEnd = time.start();
+            int newStart = time.end();
+            if (newEnd - start >= requestDuration)
+                freeTimeRanges.add(TimeRange.fromStartEnd(start, newEnd, false));
+            start = newStart;
+        }
+
+        if (TimeRange.END_OF_DAY - start >= requestDuration)
+            freeTimeRanges.add(TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, true));
+
+        return freeTimeRanges;
+    }
+    
+    public Collection < TimeRange > query(Collection < Event > events, MeetingRequest request) {
```

Comment:
I'm fine with this approach (no action required), though I'd point out you can seperate the code's concerns further by creating an auxiliary query method that checks for all users in a list, and then first calling it using the union of all required and optional attendees and, if that yields no empty slots, again with only the required ones. Something like:

Auxiliary method:
private static  Collection\<TimeRange\> query(Collection\<Event\> events, Collection\<string\> attendees);

public Collection\<TimeRange\> query(Collection\<Event\> events, MeetingRequest request) {
  Collection\<TimeRange\> ret = query(events, request.GetAttendees().addAll(request.getOptionalAttendees()));
  return !ret.isEmpty() ? ret : query(events, request.GetAttendees());
}

This isolates the concerns related to optional attendees to just a few lines, rather than intertwining it with the rest of the logic (making it harder to follow/maintain).